db/user_js_runner.go

```diff
@@ -0,0 +1,384 @@
+	// Instead, look for the name as a prefix. (See the `ottoResult` function below)
+	str := err.Error()
+	if strings.HasPrefix(str, "HTTP:") {
+		m := HttpErrRE.FindStringSubmatch(str)
+		status, _ := strconv.ParseInt(m[1], 10, 0)
+		message := m[2]
+		if status == http.StatusUnauthorized && (runner.currentDB.user == nil || runner.currentDB.user.Name() != "") {
+			status = http.StatusForbidden
+		}
+		return base.HTTPErrorf(int(status), "%s (while calling %s %q)", message, runner.kind, runner.name)
+	}
+	return &jsError{err, runner.kind, runner.name}
+}
+
+//////// DATABASE CALLBACK FUNCTION IMPLEMENTATIONS:
+
+// Implementation of JS `app.func(name, params)` function
+func (runner *userJSRunner) do_func(funcName string, params map[string]interface{}) (interface{}, error) {
+	return runner.currentDB.CallUserFunction(funcName, params, runner.mutationAllowed)
+}
+
+// Implementation of JS `app.get(docID, docType)` function
+func (runner *userJSRunner) do_get(docID string, docType *string) (interface{}, error) {
+	if err := runner.currentDB.CheckTimeout(); err != nil {
+		return nil, err
+	}
+	rev, err := runner.currentDB.GetRev(docID, "", false, nil)
+	if err != nil {
+		status, _ := base.ErrorAsHTTPStatus(err)
+		if status == http.StatusNotFound {
+			// Not-found is not an error; just return null.
+			return nil, nil
+		}
+		return nil, err
+	}
+	body, err := rev.Body()
+	if err != nil {
+		return nil, err
+	}
+	if docType != nil && body["type"] != *docType {
+		return nil, nil
+	}
+	body["id"] = docID
+	return body, nil
+}
+
+// Implementation of JS `app.graphql(name, params)` function
+func (runner *userJSRunner) do_graphql(query string, params map[string]interface{}) (interface{}, error) {
+	return runner.currentDB.UserGraphQLQuery(query, "", params, runner.mutationAllowed)
+}
+
+// Implementation of JS `app.query(name, params)` function
+func (runner *userJSRunner) do_query(queryName string, params map[string]interface{}) ([]interface{}, error) {
+
+	results, err := runner.currentDB.UserN1QLQuery(queryName, params)
+	if err != nil {
+		return nil, err
+	}
+	result := []interface{}{}
+	var row interface{}
+	for results.Next(&row) {
+		result = append(result, row)
+	}
+	if err = results.Close(); err != nil {
+		return nil, err
+	}
+	return result, err
+}
+
+// Implementation of JS `app.save(docID, body)` function
+func (runner *userJSRunner) do_save(docID string, body map[string]interface{}) error {
+	if err := runner.currentDB.CheckTimeout(); err != nil {
+		return err
+	}
+	if !runner.mutationAllowed {
+		return base.HTTPErrorf(http.StatusForbidden, "a read-only request is not allowed to mutate the database")
+	}
+
+	// TODO: Currently this is "last writer wins": get the current revision if any, and pass it
+	// to Put so that the save always succeeds.
+	rev, err := runner.currentDB.GetRev(docID, "", false, []string{})
+	if err == nil {
+		body["_rev"] = rev.RevID
+	}
```

Comment:
As written this still has the possibility that someone else updates the rev in between Get and Put.  Are we ok returning 409 to the caller of do_save in this case?

Reply:
You're right. Is there a method that will just slam a new revision without MVCC, or do I need to add a loop that retries when it gets a conflict?

Reply:
No I don't think we've got any other use cases for this type of LWW - probably easiest to just add retry loop.

Reply:
Fixed. (And added unit tests for this.)